polus-training-splinedist-plugin/src/utils.py

```diff
@@ -0,0 +1,520 @@
+        n_val = np.ceil((split_percentile/100) * num_inputs).astype('int')
+        ind_train, ind_val = index[:-n_val], index[-n_val:]
+        X_val, Y_val = [input_images[i] for i in ind_val]  , [input_labels[i] for i in ind_val] # splitting data into train and testing
+        X_trn, Y_trn = [input_images[i] for i in ind_train], [input_labels[i] for i in ind_train] 
+    
+    # Lengths
+    num_images_trained = len(X_trn)
+    num_labels_trained = len(Y_trn)
+    num_images_tested = len(X_val)
+    num_labels_tested = len(Y_val)
+    
+    # Get logs for end user
+    totalimages = num_images_trained+num_images_tested
+    logger.info("{}/{} images used for training".format(num_images_trained, totalimages))
+    logger.info("{}/{} labels used for training".format(num_labels_trained, totalimages))
+    logger.info("{}/{} images used for testing".format(num_images_tested, totalimages))
+    logger.info("{}/{} images used for testing".format(num_labels_tested, totalimages))
+
+
+    # Renamed inputs
+    del input_images
+    del input_labels
+    del num_inputs
+
+    # assertions
+    assert num_images_trained > 1, "Not Enough Training Data"
+    assert num_images_trained == num_labels_trained, "The number of images does not match the number of ground truths for training"
+    assert num_images_tested == num_labels_tested, "The number of images does not match the number of ground for testing"
+    num_trained = num_images_trained
+    num_tested = num_images_tested
+
+    del num_images_trained
+    del num_images_tested
+    del num_labels_trained
+    del num_labels_tested
+
+    # Need a list of numpy arrays to feed to SplineDist
+    array_images_trained = []
+    array_labels_trained = []
+    array_images_tested = []
+    array_labels_tested = []
+
+    # Neural network parameters
+    axis_norm = (0,1)
+    n_channel = 1 # this is based on the input data
+    
+    model_dir_name = '.'
+    model_dir_path = os.path.join(output_directory, model_dir_name)
+
+    # Get the testing data for neural network
+    logger.info("\n Starting to Load Test Data ...")
+    tenpercent_tested = np.floor(num_tested/10)
+    for i in range(num_tested):
+
+        image = X_val[i]
+        label = Y_val[i]
+        base_image = os.path.basename(str(image))
+        base_label = os.path.basename(str(label))
+        assert base_image == base_label, "{} and {} do not match".format(base_image, base_label)
+
+        # The original image
+        br_image = BioReader(image)
+        im_array = br_image[:,:,0:1,0:1,0:1]
+        im_array = im_array.reshape(br_image.shape[:2])
+        array_images_tested.append(normalize(im_array,pmin=1,pmax=99.8,axis=axis_norm))
+
+        # The corresponding label for the image
+        br_label = BioReader(label)
+        lab_array = br_label[:,:,0:1,0:1,0:1]
+        lab_array = lab_array.reshape(br_label.shape[:2])
+        array_labels_tested.append(fill_label_holes(lab_array))
```

Comment:
Don't forget to close the BioReader objects when you're done with them using `br_image.close()`/`br_label.close()`.

Reply:
Is it necessary to hold all data in memory while training?

Reply:
You should explain in the README that training modifies the labeled images. If someone is trying to segment a donut, training will only ever learn to segment a circle, so some data sets may not be well suited for splinedist. Obviously they can go read the paper to discover this, but if we can help give people warning signs in the documentation that would be best.